lib/doorkeeper/request.rb

```diff
@@ -3,6 +3,21 @@
 module Doorkeeper
   module Request
     class << self
+      def client_authentication_method(request)
+        # TODO: Should we support theoretically more than one method matching a
+        # request and then check each for authentication? Currently we only
+        # check the first that matches the request
```

Comment:
Currently we only return the first matching method for a request, not all potentially matching methods. I'm not sure if we should support that?

Reply:
Yeah, this is actually a bug. We need to collect an array of potential client credentials, and then check if we:
- had zero, then return an error when client authentication is required (`invalid_client`, 401)
- had more than one, return an error (`invalid_request`, 400, https://www.rfc-editor.org/rfc/rfc6749.html#section-2.3 but the error type comes from section 5.2)

I discovered this whilst implementing client authentication mechanisms in Hollo (an activitypub server)